nifi-mock/src/main/java/org/apache/nifi/util/MockProvenanceReporter.java

```diff
@@ -225,6 +228,49 @@ public void send(final FlowFile flowFile, final String transitUri, final String
         }
     }
 
+    @Override
+    public void upload(final FlowFile flowFile, final FileResource fileResource, final String transitUri) {
+        upload(flowFile, fileResource, transitUri, null, -1L, true);
+
+    }
+
+    @Override
+    public void upload(final FlowFile flowFile, final FileResource fileResource, final String transitUri, final long transmissionMillis) {
+        upload(flowFile, fileResource, transitUri, transmissionMillis, true);
+    }
+
+    @Override
+    public void upload(final FlowFile flowFile, final FileResource fileResource, final String transitUri, final String details, final long transmissionMillis) {
+        upload(flowFile, fileResource, transitUri, details, transmissionMillis, true);
+    }
+
+    @Override
+    public void upload(final FlowFile flowFile, final FileResource fileResource, final String transitUri, final long transmissionMillis, final boolean force) {
+        upload(flowFile, fileResource, transitUri, null, transmissionMillis, force);
+    }
+
+    @Override
+    public void upload(FlowFile flowFile, FileResource fileResource, String transitUri, String details, long transmissionMillis, boolean force) {
+        try {
+            final String enrichedDetails = StringUtils.isNotBlank(details) ? details + " " + fileResource.toString() : fileResource.toString();
+            final ProvenanceEventRecord record = build(flowFile, ProvenanceEventType.UPLOAD)
+                    .setTransitUri(transitUri)
+                    .setEventDuration(transmissionMillis)
+                    .setDetails(enrichedDetails)
+                    .build();
+            if (force) {
+                sharedSessionState.addProvenanceEvents(Collections.singleton(record));
+            } else {
+                events.add(record);
+            }
+        } catch (final Exception e) {
+            logger.error("Failed to generate Provenance Event due to " + e);
+            if (logger.isDebugEnabled()) {
+                logger.error("", e);
+            }
```

Comment:
The `due to` convention should be avoided, I also recommend avoid the conditional debug for the stack trace. Although it is present in some places, we should move away from it in general.